drivers/sqljs-driver/src/test/kotlin/com/squareup/sqldelight/drivers/sqljs/JsWorkerTransacterTest.kt

```diff
@@ -0,0 +1,187 @@
+      }
+
+      assertEquals(0, counter)
+    }
+
+    assertEquals(2, counter)
+  }
+
+  @Test fun afterCommit_does_not_run_in_nested_transaction_when_enclosing_rolls_back() = runTest { _, transacter ->
+    var counter = 0
+    transacter.transaction {
+      afterCommit { counter++ }
+      assertEquals(0, counter)
+
+      transaction {
+        afterCommit { counter++ }
+      }
+
+      rollback()
+    }
+
+    assertEquals(0, counter)
+  }
+
+  @Test fun afterCommit_does_not_run_in_nested_transaction_when_nested_rolls_back() = runTest { _, transacter ->
+    var counter = 0
+    transacter.transaction {
+      afterCommit { counter++ }
+      assertEquals(0, counter)
+
+      transaction {
+        afterCommit { counter++ }
+        rollback()
+      }
+
+      throw AssertionError()
+    }
+
+    assertEquals(0, counter)
+  }
+
+  @Test fun afterRollback_no_ops_if_the_transaction_never_rolls_back() = runTest { _, transacter ->
+    var counter = 0
+    transacter.transaction {
+      afterRollback { counter++ }
+    }
+
+    assertEquals(0, counter)
+  }
+
+  @Test fun afterRollback_runs_after_a_rollback_occurs() = runTest { _, transacter ->
+    var counter = 0
+    transacter.transaction {
+      afterRollback { counter++ }
+      rollback()
+    }
+
+    assertEquals(1, counter)
+  }
+
+  @Test fun afterRollback_runs_after_an_inner_transaction_rolls_back() = runTest { _, transacter ->
+    var counter = 0
+    transacter.transaction {
+      afterRollback { counter++ }
+      transaction {
+        rollback()
+      }
+      throw AssertionError()
+    }
+
+    assertEquals(1, counter)
+  }
+
+  @Test fun afterRollback_runs_in_an_inner_transaction_when_the_outer_transaction_rolls_back() = runTest { _, transacter ->
+    var counter = 0
+    transacter.transaction {
+      transaction {
+        afterRollback { counter++ }
+      }
+      rollback()
+    }
+
+    assertEquals(1, counter)
+  }
+
+  @Test fun transactions_close_themselves_out_properly() = runTest { _, transacter ->
+    var counter = 0
+    transacter.transaction {
+      afterCommit { counter++ }
+    }
+
+    transacter.transaction {
+      afterCommit { counter++ }
+    }
+
+    assertEquals(2, counter)
+  }
+
+  @Test fun setting_no_enclosing_fails_if_there_is_a_currently_running_transaction() = runTest { _, transacter ->
+    transacter.transaction(noEnclosing = true) {
+      assertFailsWith<IllegalStateException> {
+        transacter.transaction(noEnclosing = true) {
+          throw AssertionError()
+        }
+      }
+    }
+  }
+
+  @Test fun an_exception_thrown_in_postRollback_function_is_combined_with_the_exception_in_the_main_body() = runTest { _, transacter ->
+    class ExceptionA : RuntimeException()
+    class ExceptionB : RuntimeException()
+    try {
+      transacter.transaction {
+        afterRollback {
+          throw ExceptionA()
+        }
+        throw ExceptionB()
+      }
+      fail("Should have thrown!")
```

Comment:
Would prefer `assertFailsWith` since currently this will throw an `AssertionError` that will be caught which will then fail only because we're doing the checks below. `assertFailsWith` is slightly safer.

Reply:
I don't think that's applicable here because the `fail` call is never reached since `ExceptionB()` is thrown. The actual exception that is caught is a `RollbackException` which is private in `runtime`/`runtime-async` and this test exists just to test that both exceptions A and B show up in the `RollbackException` message.

This was just copied from the existing `JsTransacterTest`s which were adapted from the `driver-test` module. I'm not actually sure this test is even needed to be honest.

Reply:
The point of `assertFailsWith` is to replace the `try { statement(); fail() } catch { }` pattern. It validates that the body of the lambda never completes successfully. I will can handle replacing this and the upstream usage you cite with it since I'll be in the codebase here shortly.

Reply:
Ah, I see how that works now. Thanks